### ISSUE TITLE
slow search with big tables

### ISSUE THREAD

**Issue by TZanke** [deitch/jstree-grid#208]
We are using jstree-grid to show a table with 570 rows and 6 columns of data. This takes a lot of time when the search matches a great amount of the rows.
If i disable the grid, the search is very quick, so i looked into the problem.

Using the Chrome Profiler i was able to detect the source of the slowdown. This foreach:
https://github.com/deitch/jstree-grid/blob/b9dd70a64192f151755bd19539dd36614109eeab/jstreegrid.js#L521

What do you think? What could i do to get rid of the extra waiting time?

**Comment by deitch** [comment]
Let's try to reason it through. When it receives a `search.jstree` event, it has already run the search and passes it the correctly matching nodes. 

Ignoring the `search.som` option, it then calls `findDataCell()` for each of them and adds to all the class `SEARCHCLASS`.

So here's the question. Which of these is slow:

1. `data.nodes.filter('.jstree-node')`
2. `.each(function(i,node) {...})`
3. `findDataCell(grid, node.id)`
4. `addClass(SEARCHCLASS)`

I highly doubt it is `addClass` and suspect `findDataCell`, but not sure. Can you profile the individual parts, or perhaps modify to run innocent steps and isolate where it is coming from?

**Comment by TZanke** [comment]
Enabling one after another:

1. takes some milliseconds
2. only some milliseconds
3. needs a complete second to get processed
3. does **not** add a measurable amount of time to 3.

So it looks like the findDataCell bottlenecks the selection process.

The each calls findDataCell 554 times.

**Comment by deitch** [comment]
Well, that makes sense, and is what I expected, but was worth checking before going down the rabbit hole.

> 3. needs a complete second to get processed

A second for each of the 554 times? Or a second for all 554 cumulatively, which is 1.8ms each time? (I sure hope the second....)

**Comment by TZanke** [comment]
yes, 1000 ms / 554.

**Comment by deitch** [comment]
So `findDataCell()` itself is reasonably speedy, but calling it in a large `each()` loop gets messy. Need to find a better algorithm for it. Thinking....

**Comment by deitch** [comment]
Just to verify for the second point. Did you change the code so `findDataCell()` just did something silly - like `console.log("I AM HERE");` - and call it, so we get the `findDataCell()` called 554 times but no performance impact of what is inside the function? I want to be sure that it is the _contents_ of `findDataCell()` that take 1.8ms each, and not necessarily the calling of _any_ function (pushing it onto the stack, saving state, etc.) that is taking 1.8ms each.

**Comment by deitch** [comment]
There is a branch here https://github.com/deitch/jstree-grid/tree/faster-findGridCell for it.

This doesn't optimize anything yet, just makes `findDataCell()` able to take a single node, id or list of nodes. But that does give us one place to optimize it.

Please verify that we see the same performance hit here, then we can optimize `findDataCell()`.

**Comment by TZanke** [comment]
Yes, the _content_ of findDataCell is slow, _not_ the stack etc.

I tested your new branch, it only takes 800ms but also brings some errors:

`jstreegrid.js:503 Uncaught TypeError: findDataCell(...).removeClass is not a function`

so i think this is because of undone work because of canceled operations.

**Comment by deitch** [comment]
> it only takes 800ms 

I didn't expect anything other than a slight speedup, because I pulled a little bit out of the loop. The main point was to have it in one place.

> also brings some errors

I figured it would, given that it was quick and untested. But if you tell me how that got triggered, it would help. :-)

Now we can focus on actually optimizing `findDataCell()`.

**Comment by deitch** [comment]
OK, never mind on the error, got that. Should have returned `$()` instead of `[]`

**Comment by deitch** [comment]
@TZanke : do you hace a sample large data set? It would be most helpful to use for testing.

**Comment by deitch** [comment]
Forget the data set. Just generated one in the test pushed.

**Comment by deitch** [comment]
And now the test shows time. I have 1,000 rows, and using the current logic, it still only takes ~60ms to do the search.

Look at https://github.com/deitch/jstree-grid/blob/faster-findGridCell/tests/components/large-dataset-test.html and see if you can figure out why you see 800ms and I get ~65ms? What is different about my use case?

**Comment by TZanke** [comment]
I switched to you branch and verified the test:

searching for "child" takes:

- Chrome: Time for search: 469 milliseconds
- Firefox: Time for search: 493 milliseconds 

Another problem: Opening the root node by clicking on the little arrow already takes 1000ms in large-dataset.test.html

**Comment by deitch** [comment]
That makes sense. I was searching for a low-match set. 

I saw the root node, not sure how much there is to do. It might be as much a tree problem as the grid. Anything in profiler?

**Comment by deitch** [comment]
Take a look at the latest push. I am showing 22ms (!) in Chrome. Opening a PR now...

**Comment by deitch** [comment]
Once you agree that it fixes it, merging in.

**Comment by TZanke** [comment]
i pulled you code right now.

Now it shows 26ms, **BUT it is still slow!**

![image](https://cloud.githubusercontent.com/assets/1613368/22105172/1f7cddfe-de43-11e6-9de3-378c1f1417f5.png)

**Comment by deitch** [comment]
I don't get it. How can it be 26ms but slow?

**Comment by TZanke** [comment]
It looks like the time measurement does not capture the slow part of the code?

**Comment by deitch** [comment]
OK, but you have to admit that is strange. Look at the code:

```js
				var grid = this.gridWrapper, that = this, nodesToShow, startTime = new Date().getTime(), endTime;
				this.holdingCells = {};
				if(data.nodes.length) {
					if(this._data.search.som) {
						// create the list of nodes we want to look at
						if(this._data.search.smc) {
							nodesToShow = data.nodes.add(data.nodes.find('.jstree-node'));
						}
						nodesToShow = (nodesToShow || data.nodes).add(data.nodes.parentsUntil(".jstree"));

						// hide all of the grid cells
						grid.find('div.jstree-grid-cell-regular').hide();
						// show only those that match
						nodesToShow.filter(".jstree-node").each(function (i,node) {
							var id = node.id;
							if (id) {
								that._prepare_grid(node);
								findDataCell(grid,id).show();
							}
						});
					}

					findDataCell(grid,data.nodes.filter(".jstree-node")).addClass(SEARCHCLASS);
          endTime = new Date().getTime();
          this.element.trigger("search-complete.jstree-grid", [{time:endTime-startTime}]);
				}
				return true;

```

It takes the start time from the beginning, end time from the end, and sends the event at the very end with the diff. Plus, I saw your ~800ms when I ran it before the change, and after making the changes, it slid down to 26ms. 

Where exactly are you seeing the slowness in profile?

**Comment by TZanke** [comment]
I am no JS profiling expert, but i try to help as best i could.

![image](https://cloud.githubusercontent.com/assets/1613368/22113110/88b34660-de65-11e6-89a1-e1896cafc480.png)

So elemData.handle is called several times (for each found row i think) and splits up in two or three _prepare_grid each time elemData.handle is called.

**Comment by TZanke** [comment]
The _cumulated_ most time consuming called function is  [querySelectorAll](https://developer.mozilla.org/de/docs/Web/API/Element/querySelectorAll) .

**Comment by deitch** [comment]
> So elemData.handle is called several times (for each found row i think) and splits up into multiple (1-n times) _prepare_grid each time elemData.handle is called.

Yes, but `_prepare_grid` is not called from within the search. I just checked again, even putting a breakpoint there.

Your search still is taking hundreds of ms?

**Comment by TZanke** [comment]
Look at the Stacktrace, search calls _redraw_node_ of jstreegrid which calls __prepare_grid_

![image](https://cloud.githubusercontent.com/assets/1613368/22140918/aa8205f8-deef-11e6-8562-b3e837dc72e2.png)


![image](https://cloud.githubusercontent.com/assets/1613368/22141036/4469e564-def0-11e6-9ddb-759fa4b3e8e0.png)

**Comment by deitch** [comment]
> search calls redraw_node of jstreegrid which calls _prepare_grid

Yes, `redraw_node` calls `_prepare_grid`, which is a reasonably expensive operation with all of its calculations. It probably could afford to be optimized in and of itself, but that won't matter much here.

But when I run search, it does _not_ call `redraw_node`. That probably is why mine completes so quickly. 

Are you using the test file from the branch? I just ran profiler again, and I am seeing 30-40ms for the search and no calls to `redraw_node`. E.g. searching for `child2` gave me 39.9ms in `querySelectorAll`.

**Comment by deitch** [comment]
Any chance you have `show_only_matches` set to `true`? That might trigger redrawing with search.

**Comment by TZanke** [comment]
show_only_matches is False, i just use your branch. I dont know why i am experiencing other behavior.

if show_only_matches is True, it takes half a minute and one CPU core is used 100%.

**Comment by deitch** [comment]
> if show_only_matches is True, it takes half a minute and one CPU core is used 100%.

That isn't too surprising. `show_only_matches` is an expensive operation for a large tree, even more so with the grid.

> show_only_matches is False, i just use your branch. I dont know why i am experiencing other behavior

Neither do I. `redraw_node` is *not* in the path for search. 

Oh wait a second. Are you doing the search _before_ or _after_ opening the root node?

**Comment by TZanke** [comment]
root node is closed.

**Comment by TZanke** [comment]
ok yes, if i first open root node, and then doing the search its very quick.

So it looks like opening the root node is slow.

**Comment by deitch** [comment]
Aha! It is not the _search_ that is slow, it is the _open_ of the root node that is slow. If you opened the root node first and then did the search, search itself would complete in a few tens of milliseconds even on that 1,000-node tree.

That is why the search complete event is showing so little time. It is only called once jstree is done opening its nodes.

I am going to merge in the PR - we definitely have an order+ of magnitude improvement in search on trees when open - but let's do another experiment.

Can you profile the time to:

* open the root node without grid
* open the root node with grid

No searching at all. Let's see how much time grid adds to the process.

**Comment by TZanke** [comment]
First part comes from jstree, marked part is added if i enable grid for this table.
![image](https://cloud.githubusercontent.com/assets/1613368/22146763/0f6c70a0-df06-11e6-96d7-df5e85c1a5bd.png)

(used my tree data)

I am using a method which opens nodes automatically. So the site "hangs" after loading till grid is finished. 

After waiting for the grid the user now can search quickly after all. **Thx for this**.

**Comment by deitch** [comment]
I am having a hard time seeing the graph. How much time is the tree with grid disabled, how much with grid enabled?

**Comment by TZanke** [comment]
jstreegrid adds ~2500ms to the execution in my tree in comparision to jstree only.

**Comment by deitch** [comment]
That _is_ a lot. Is that using the branch I just merged in with the 1,000-node test? Or is it your specific use case?

**Comment by TZanke** [comment]
i pulled the master, my data and your 1000 row test: Grid adds 2500ms

**Comment by deitch** [comment]
Did you reports show how much the large jstree is without the grid?

**Comment by TZanke** [comment]
150ms without grid, it feels instant

**Comment by deitch** [comment]
Yeah, that is an issue, and the last figure proves the issue is in the grid.

I am closing this issue out, since we resolved the search problem. Can you open a new issue with just those last numbers (150ms tree alone with the large dataset sample from `tests/`, another 2500ms for grid)? One issue per issue.

Thanks again for the sleuthing.
